### PR TITLE
Add support for recursive copy of input and output folders

### PR DESCRIPTION
**Pull request by jbingham**
Users have asked for this feature. 

Also fixed some edge case bugs, and began some cleanup to be finished in a separate PR.

### CODE REVIEW

File: src/main/java/com/google/cloud/genomics/dockerflow/task/Task.java [hunk 1]
```diff
@@ -390,6 +388,146 @@ private void resolveLocalCopies() {
...
+                "${" + p.getName() + "}", 
+                DockerflowConstants.DEFAULT_MOUNT_POINT + 
+                p.getLocalCopy().getPath() + "\""));
+
+        // Turn it into an env var
+        p.setLocalCopy(null);
+        p.setType(null);
+                
+        // Move output folders to input parameters
+        // because outputs can only be files
```
Comment (mbookman) [inline src/main/java/com/google/cloud/genomics/dockerflow/task/Task.java]: Can you expand this comment further? I don't understand it.

File: src/main/java/com/google/cloud/genomics/dockerflow/task/Task.java [hunk 2]
```diff
@@ -390,6 +388,146 @@ private void resolveLocalCopies() {
...
+        if (!inputs.contains(p.getName())) {
+          defn.getOutputParameters().remove(p);
+          if (defn.getInputParameters() == null) {
+            defn.setInputParameters(new LinkedHashSet<Param>());
+          }
+          defn.getInputParameters().add(p);
+          
+          // move output folder args to input args
```
Comment (mbookman) [inline src/main/java/com/google/cloud/genomics/dockerflow/task/Task.java]: s/move/Move

File: src/main/java/com/google/cloud/genomics/dockerflow/task/Task.java [hunk 3]
```diff
@@ -390,6 +388,146 @@ private void resolveLocalCopies() {
...
+              args.setInputs(new LinkedHashMap<String,String>());
+            }
+            args.getInputs().put(p.getName(), args.getOutputs().get(p.getName()));
+            args.getOutputs().remove(p.getName());
+          }
+        }
+      }
+    }
+    
+    // Pipelines API chokes on empty output parameter set
```
Comment (mbookman) [inline src/main/java/com/google/cloud/genomics/dockerflow/task/Task.java]: s/chokes on empty output parameter set/rejects empty parameter values.

File: src/main/java/com/google/cloud/genomics/dockerflow/task/Task.java [hunk 4]
```diff
@@ -390,6 +388,146 @@ private void resolveLocalCopies() {
...
+            "    2>&1 echo \"Recursive delocalization failed.\"\n" +
+            "    exit 1\n" +
+            "  fi\n" +
+            "done\n\n");
+      }
+    }
+    if (!gcsPaths.isEmpty()) {
+      defn.getDocker().setCmd(
+          "\n" +
+          "# Install gsutil\n" +
```
Comment (mbookman) [inline src/main/java/com/google/cloud/genomics/dockerflow/task/Task.java]: I think this whole block of installation code should be pulled out to a separate string constant such that this block can be something more concise like:

    defn.getDocker().setCmd(
         "\n" +
         "# Install gsutil\n" +
         INSTALL_GCLOUD_CMD + "\n" +
         "# Copy inputs\n" +
         copyInputs.toString() + "\n" +
         "# Run user script\n" +
         defn.getDocker().getCmd() + "\n" +
         "# Copy outputs\n" +
         copyOutputs.toString());

**Comment by jbingham** [comment]
Thanks for the suggestions. Incorporated.

File: src/main/java/com/google/cloud/genomics/dockerflow/task/Task.java [hunk 5]
```diff
@@ -390,6 +388,151 @@ private void resolveLocalCopies() {
...
+
+    // Edit the Docker command to copy the folders
+    StringBuilder copyInputs = new StringBuilder();
+    StringBuilder copyOutputs = new StringBuilder();
+    
+    for (String name : gcsPaths.keySet()) {
+      // Copy input folder
+      if (inputs.contains(name)) {
+        copyInputs.append(
+            "mkdir -p " +
```
Comment (mbookman) [inline src/main/java/com/google/cloud/genomics/dockerflow/task/Task.java]: Sorry - I had one more comment that seems to have gotten lost.

I suggest reformatting this as:

    copyInputs.append(
        String.format(
            "mkdir -p %s/%s\n",
            DockerflowConstants.DEFAULT_MOUNT_POINT,
            localPaths.get(name)) +
        String.format(
            "for ((i = 0; i < 3; i++)); do\n" +
            "  if gsutil -m rsync -r %s/ %s/%s; then\n" +
            "    break\n" +
            "  elif ((i == 2)); then\n" +
            "    2>&1 echo \"Recursive localization failed.\"\n" +
            "    exit 1\n" +
            "  fi\n",
            gcsPaths.get(name),
            DockerflowConstants.DEFAULT_MOUNT_POINT,
            localPaths.get(name)))

to make it more clear what the generated shell script will look like.

Comment (jbingham) [inline src/main/java/com/google/cloud/genomics/dockerflow/task/Task.java]: Agreed. Done.

**Pull request merged**